pkg/controllers/provisioning/v1alpha1/reallocation/suite_test.go

```diff
@@ -0,0 +1,170 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reallocation
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/Pallinder/go-randomdata"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/cloudprovider/fake"
+	"github.com/awslabs/karpenter/pkg/test"
+	"github.com/awslabs/karpenter/pkg/test/environment"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	. "github.com/awslabs/karpenter/pkg/test/expectations"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest/printer"
+)
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecsWithDefaultAndCustomReporters(t,
+		"Provisioner",
+		[]Reporter{printer.NewlineReporter{}})
+}
+
+var controller *Controller
+var env environment.Environment = environment.NewLocal(func(e *environment.Local) {
+	controller = NewController(
+		e.Manager.GetClient(),
+		corev1.NewForConfigOrDie(e.Manager.GetConfig()),
+		fake.NewFactory(cloudprovider.Options{}),
+	)
+	e.Manager.Register(controller)
+})
+
+var _ = BeforeSuite(func() {
+	Expect(env.Start()).To(Succeed(), "Failed to start environment")
+})
+
+var _ = AfterSuite(func() {
+	Expect(env.Stop()).To(Succeed(), "Failed to stop environment")
+})
+
+var _ = Describe("Reallocation", func() {
+	var ns *environment.Namespace
+
+	BeforeEach(func() {
+		var err error
+		ns, err = env.NewNamespace()
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		ExpectCleanedUp(ns.Client)
+	})
+
+	Context("Reconciliation", func() {
+		It("should label nodes for underutilized", func() {
+			p := &v1alpha1.Provisioner{ObjectMeta: metav1.ObjectMeta{Name: strings.ToLower(randomdata.SillyName()), Namespace: ns.Name}}
+			node := test.NodeWith(test.NodeOptions{
+				Labels: map[string]string{
+					v1alpha1.ProvisionerNameLabelKey:      p.Name,
+					v1alpha1.ProvisionerNamespaceLabelKey: p.Namespace,
+				},
+			})
+			ExpectCreatedWithStatus(ns.Client, node)
+
+			ExpectCreated(ns.Client, p)
+			ExpectEventuallyReconciled(ns.Client, p)
+
+			Expect(ns.Client.Get(context.Background(), client.ObjectKey{Name: node.Name}, node)).To(Succeed())
+			Expect(node.Labels).To(HaveKeyWithValue(v1alpha1.ProvisionerPhaseLabel, v1alpha1.ProvisionerUnderutilizedPhase))
+			Expect(node.Annotations).To(HaveKey(v1alpha1.ProvisionerTTLKey))
+		})
+
+		It("should remove labels from utilized nodes", func() {
```

Comment:
I think it would be good to add a test that cycles nodes thru the states a few times? Like say it starts out underutilized, then goes to utilized, then goes back to underutilized?

Reply:
Good idea. I think an e2e test would be good too.

Reply:
Punting on this for a follow-up PR.